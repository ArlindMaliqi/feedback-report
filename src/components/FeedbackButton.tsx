import React from "react";
import { useFeedback } from "../hooks/useFeedback";
import { useTheme } from "../hooks/useTheme";

/**
 * Props for the FeedbackButton component
 */
interface FeedbackButtonProps {
  /** Position of the button on screen */
  position?: "bottom-right" | "bottom-left" | "top-right" | "top-left";
  /** Size of the button in pixels */
  size?: number;
  /** Background color of the button */
  backgroundColor?: string;
  /** Text color of the button */
  textColor?: string;
  /** Text to display inside the button */
  text?: string;
  /** Tooltip text (default: "Send Feedback") */
  tooltip?: string;
}

/**
 * A floating feedback button that opens the feedback modal when clicked
 *
 * This component renders a fixed-position circular button that users can click
 * to open the feedback modal. It automatically adapts to light/dark themes.
 *
 * @param props - Component props
 * @param props.position - Where to position the button (default: "bottom-right")
 * @param props.size - Button size in pixels (default: 60)
 * @param props.backgroundColor - Button background color (default: automatic based on theme)
 * @param props.textColor - Button text color (default: "white")
 * @param props.text - Button text content (default: "?")
 * @param props.tooltip - Tooltip text (default: "Send Feedback")
 */
export const FeedbackButton: React.FC<FeedbackButtonProps> = ({
  position = "bottom-right",
  size = 60,
  backgroundColor,
  textColor,
  text = "?",
  tooltip = "Send Feedback",
}) => {
  const { openModal } = useFeedback();
  const { theme } = useTheme();

  // Use theme-aware colors if not explicitly provided
  const bgColor = backgroundColor || (theme === "dark" ? "#3b82f6" : "#007bff");
  const txtColor = textColor || "white";

  const getPositionStyles = (): Record<string, string> => {
    const [vertical, horizontal] = position.split("-") as [string, string];
    return {
      [vertical]: "20px",
      [horizontal]: "20px",
    };
  };

  return (
    <button
      onClick={openModal}
      aria-label={tooltip}
      title={tooltip}
      style={{
        position: "fixed",
        ...getPositionStyles(),
        backgroundColor: bgColor,
        color: txtColor,
        border: "none",
        borderRadius: "50%",
        width: `${size}px`,
        height: `${size}px`,
        fontSize: `${size * 0.4}px`,
        fontWeight: "bold",
        boxShadow: "0 4px 12px rgba(0, 0, 0, 0.15)",
        cursor: "pointer",
        zIndex: 999,
        transition: "all 0.2s ease",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
      }}
      onMouseEnter={(e) => {
        e.currentTarget.style.transform = "scale(1.1)";
        e.currentTarget.style.boxShadow = "0 6px 16px rgba(0, 0, 0, 0.2)";
      }}
      onMouseLeave={(e) => {
        e.currentTarget.style.transform = "scale(1)";
        e.currentTarget.style.boxShadow = "0 4px 12px rgba(0, 0, 0, 0.15)";
      }}
    >
      {text}
    </button>
  );
};

export default FeedbackButton;
