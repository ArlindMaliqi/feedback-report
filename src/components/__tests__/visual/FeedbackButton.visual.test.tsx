import React from 'react';
import { renderWithFeedback } from '../../../testing/testUtils';
import { FeedbackButton } from '../../FeedbackButton';

describe('FeedbackButton Visual Tests', () => {
  it('renders correctly in light theme', () => {
    const { container } = renderWithFeedback(<FeedbackButton />, {
      config: { theme: 'light' }
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly in different positions', () => {
    const positions = ['bottom-right', 'bottom-left', 'top-right', 'top-left'] as const;
    
    positions.forEach(position => {
      const { container } = renderWithFeedback(
        <FeedbackButton position={position} />
      );
      
      expect(container.firstChild).toMatchSnapshot(`position-${position}`);
    });
  });
});
