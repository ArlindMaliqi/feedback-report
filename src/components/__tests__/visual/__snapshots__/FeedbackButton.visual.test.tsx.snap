// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackButton Visual Tests renders correctly in different positions: position-bottom-left 1`] = `
<button
  aria-label="Feedback"
  style="position: fixed; bottom: 20px; left: 20px; background-color: rgb(0, 123, 255); color: white; border-radius: 50%; cursor: pointer; box-shadow: 0 2px 10px rgba(0,0,0,0.2); z-index: 1000; display: flex; align-items: center; justify-content: center; width: 56px; height: 56px; font-size: 16px;"
  title="Feedback"
>
  ðŸ’¬
</button>
`;

exports[`FeedbackButton Visual Tests renders correctly in different positions: position-bottom-right 1`] = `
<button
  aria-label="Feedback"
  style="position: fixed; bottom: 20px; right: 20px; background-color: rgb(0, 123, 255); color: white; border-radius: 50%; cursor: pointer; box-shadow: 0 2px 10px rgba(0,0,0,0.2); z-index: 1000; display: flex; align-items: center; justify-content: center; width: 56px; height: 56px; font-size: 16px;"
  title="Feedback"
>
  ðŸ’¬
</button>
`;

exports[`FeedbackButton Visual Tests renders correctly in different positions: position-top-left 1`] = `
<button
  aria-label="Feedback"
  style="position: fixed; top: 20px; left: 20px; background-color: rgb(0, 123, 255); color: white; border-radius: 50%; cursor: pointer; box-shadow: 0 2px 10px rgba(0,0,0,0.2); z-index: 1000; display: flex; align-items: center; justify-content: center; width: 56px; height: 56px; font-size: 16px;"
  title="Feedback"
>
  ðŸ’¬
</button>
`;

exports[`FeedbackButton Visual Tests renders correctly in different positions: position-top-right 1`] = `
<button
  aria-label="Feedback"
  style="position: fixed; top: 20px; right: 20px; background-color: rgb(0, 123, 255); color: white; border-radius: 50%; cursor: pointer; box-shadow: 0 2px 10px rgba(0,0,0,0.2); z-index: 1000; display: flex; align-items: center; justify-content: center; width: 56px; height: 56px; font-size: 16px;"
  title="Feedback"
>
  ðŸ’¬
</button>
`;

exports[`FeedbackButton Visual Tests renders correctly in light theme 1`] = `
<button
  aria-label="Feedback"
  style="position: fixed; bottom: 20px; right: 20px; background-color: rgb(0, 123, 255); color: white; border-radius: 50%; cursor: pointer; box-shadow: 0 2px 10px rgba(0,0,0,0.2); z-index: 1000; display: flex; align-items: center; justify-content: center; width: 56px; height: 56px; font-size: 16px;"
  title="Feedback"
>
  ðŸ’¬
</button>
`;
